keys/ecc.go

```diff
@@ -26,6 +28,67 @@ var _ ECC = NoECC{}
 func (_ NoECC) AddECC(input []byte) []byte            { return input }
 func (_ NoECC) CheckECC(input []byte) ([]byte, error) { return input, nil }
 
+// CRC16 does the ieee crc16 polynomial check
+type CRC16 struct {
+	Poly  uint16
+	table *crc16.Table
+}
+
+var _ ECC = &CRC16{}
+
+const crc16Size = 2
+
+func NewIBMCRC16() *CRC16 {
+	return &CRC16{Poly: crc16.IBM}
+}
+
+func NewSCSICRC16() *CRC16 {
+	return &CRC16{Poly: crc16.SCSI}
+}
+
+func NewCCITTCRC16() *CRC16 {
+	return &CRC16{Poly: crc16.CCITT}
+}
+
+func (c *CRC16) AddECC(input []byte) []byte {
+	table := c.getTable()
+
+	// get crc and convert to some bytes...
+	crc := crc16.Checksum(input, table)
+	check := make([]byte, crc16Size)
+	binary.BigEndian.PutUint16(check, crc)
+
+	// append it to the input
+	output := append(input, check...)
+	return output
+}
+
+func (c *CRC16) CheckECC(input []byte) ([]byte, error) {
+	table := c.getTable()
+
+	if len(input) <= crc16Size {
+		return nil, errors.New("input too short, no checksum present")
```

Comment:
We could move this out of this function and make a var to avoid allocations on a potential hot path ie every invocation of this function ie
```go
var errInputTooShortNoChecksum = errors.New("input too short, no checksum present")
```
and then in that block:
```go
if len(input) <= crc16Size {
   return nil, errInputTooShortNoChecksum
}
```